### DISCUSSION TITLE
operator void() and Result-like types

### DISCUSSION THREAD

**Discussion by dylanrafael05** [dotnet/csharplang#6818] [Language Ideas]
This is gonna be a short one, but it might be useful to some people, particularly when discussing the concept of a `Result` type.

Currently, non-void methods must return a value explicitly. This makes sense 99% of the time, but occassionally creates verbosity:
```cs
public static MyResultType SomeAction() => MyResultType.Success;
```

Although this is a minor problem, I could see it being a head-ache in large-scale usages of such void-like types. For this reason, I would like to suggest the inclusion of a `operator void()`, which would handle creating a "void-like" result:

```cs
public class MyResultType
{
     public static MyResultType operator void() => MyResultType.Success;
     
     // . . . implementation elided . . . 
}

// Later usage: //
public static MyResultType SomeAction() {} //Ok! equivalent to returning MyResultType.Success
public static MyResultType ComplexAction(int a)
{
     if(a == 0) return; //Same as above; implicitly returns the result of the return type's operator void()
     Console.WriteLine("the input is not zero! wow!");

     if(a == 1) return MyResultType.Failure("1 is not allowed"); //Note that normal non-void returns can be mixed in with void returns in this case.

     //Close of function still returns the result of operator void() if a return is needed.
}
```

Note that the return type of `operator void()` would have to be the same as the enclosing type.
Alternate syntax for the declaration could be `operator MyResultType(void)` or `operator MyResultType()`, although that is less C#-like.

**Comment by HaloFour** [comment]
IMO it's a bad idea for a function to have a "default" return value.  A behavior similar to this, in that functions had implicit return values, existed in VB6, and it was a ripe source of potential bugs as it effectively short-circuited any compiler analysis that would guarantee that all code paths would return a value.

**Reply by dylanrafael05** [reply]
The resultant returns would be treated as if they were void returns though, meaning that the same rules for determining whether a block of code returned would apply, except that the function scope would always return implicitly at its close.

**Reply by HaloFour** [reply]
That is the problem, the compiler would never ensure that a value is returned, it would just return the default.  It would be trivial for a missed branch to just "fall out" of the method with that default value incorrectly, and this was a legit source of bugs in VB6.
